skimage/_shared/utils.py

```diff
@@ -450,3 +450,31 @@ def _fix_ndimage_mode(mode):
     if NumpyVersion(scipy.__version__) >= '1.6.0':
         mode = grid_modes.get(mode, mode)
     return mode
+
+
+def _float_type(image):
+    """Return an appropriate floating-point dtype for a given image.
+
+    float32, float64, complex64, complex128 are preserved.
+    float16 is promoted to float32.
+    complex256 is demoted to complex128.
+    Other types are cast to float64.
+
+    Paramters
+    ---------
+    image : ndarray
+        Input image.
+
+    Retruns
+    -------
+    float_type : dtype
+        Floating-point dtype for the image.
+    """
+    char = image.dtype.char
+    if char == 'e':
+        return np.float32
+    elif char == 'G':
+        return np.complex128
+    elif char not in 'dfDF':
+        return np.float64
+    return image.dtype
```

Comment:
@FirefoxMetzger I don't think your code is equivalent? There's a `return image.dtype` in Greg's code that's not accounted for in your suggestion, afaict.

However, I agree that a dict much more readable. I prefer functions to have just one or two return statements, rather than a cascade of them.

Reply:
Yes. if I understand the intention of the function correctly, then `image.dtype` is actually a bug. The description says that "everything else will be converted to FP64", but it actually isn't.

Either the docstring should get an update, or the behavior should change. See the comment below.

Reply:
I agree readability/style can be improved and am fine with a dictionary based approach.

The last `return image.dtype` can only be reached for characters 'dfDF' which are float32, float64, complex64 and complex128. These are the types are currently was not modifying, so I don't think it is a bug. Clearly the fact that we are having this discussion means clarity can be improved!

I originally had a much simpler function that was just `return np.promote_types(image.dtype, np.float32)` but that one converts smaller size ints to float32, but larger ints to float64 and also did not deal with downcasting float128/complex256.

Reply:
dict/elif is fine with me. If we go for an `elif` version, I would suggest switching the logic of the last elif to `char in 'dDfF'` and introducing a final `else` as in 

```
if char == 'e':
    new_dtype = np.float32
elif ...
    new_dtype = ...
else:
    new_dtype = np.float64

return new_dtype
```

since that gives the function a single return statement and makes the flow more intuitive (to me anyway).